// pending-xhr-puppeteer/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

var PendingXHR = /** @class */ (function () {
    function PendingXHR(page) {
        var _this = this;
        this.promisees = [];
        this.page = page;
        this.resourceType = 'xhr';
        this.pendingXhrs = new Set();
        this.finishedWithSuccessXhrs = new Set();
        this.finishedWithErrorsXhrs = new Set();
        this.requestListener = function (request) {
            if (request.resourceType() === _this.resourceType) {
                _this.pendingXhrs.add(request);
                _this.promisees.push(new Promise(function (resolve) {
                    request.pendingXhrResolver = resolve;
                }));
            }
        };
        this.requestFailedListener = function (request) {
            if (request.resourceType() === _this.resourceType) {
                _this.pendingXhrs.delete(request);
                _this.finishedWithErrorsXhrs.add(request);
                if (request.pendingXhrResolver) {
                    request.pendingXhrResolver();
                }
                delete request.pendingXhrResolver;
            }
        };
        this.requestFinishedListener = function (request) {
            if (request.resourceType() === _this.resourceType) {
                _this.pendingXhrs.delete(request);
                _this.finishedWithSuccessXhrs.add(request);
                if (request.pendingXhrResolver) {
                    request.pendingXhrResolver();
                }
                delete request.pendingXhrResolver;
            }
        };
        page.on('request', this.requestListener);
        page.on('requestfailed', this.requestFailedListener);
        page.on('requestfinished', this.requestFinishedListener);
    }
    PendingXHR.prototype.removePageListeners = function () {
        this.page.removeListener('request', this.requestListener);
        this.page.removeListener('requestfailed', this.requestFailedListener);
        this.page.removeListener('requestfinished', this.requestFinishedListener);
    };
    PendingXHR.prototype.waitForAllXhrFinished = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.pendingXhrCount() === 0) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, Promise.all(this.promisees)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    PendingXHR.prototype.waitOnceForAllXhrFinished = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.waitForAllXhrFinished()];
                    case 1:
                        _a.sent();
                        this.removePageListeners();
                        return [2 /*return*/];
                }
            });
        });
    };
    PendingXHR.prototype.pendingXhrCount = function () {
        return this.pendingXhrs.size;
    };
    return PendingXHR;
}());

exports.PendingXHR = PendingXHR;
